test/ShallowWrapper-spec.js

```diff
@@ -2455,6 +2455,575 @@ describe('shallow', () => {
+            return <div>foo</div>;
+          }
+        }
+        shallow(<Foo />, { lifecycleExperimental: true });
+        expect(spy.args).to.deep.equal([
+          ['componentWillMount'],
+          ['componentDidMount'],
+        ]);
+      });
+
+      it('should be batching updates', () => {
+        const spy = sinon.spy();
+        class Foo extends React.Component {
+          constructor(props) {
+            super(props);
+            this.state = {
+              count: 0,
+            };
+          }
+          componentWillMount() {
+            this.setState({ count: this.state.count + 1 });
+            this.setState({ count: this.state.count + 1 });
+          }
+          componentDidMount() {
+            this.setState({ count: this.state.count + 1 });
+            this.setState({ count: this.state.count + 1 });
+          }
+          render() {
+            spy();
+            return <div>{this.state.count}</div>;
+          }
+        }
+        const result = shallow(<Foo />, { lifecycleExperimental: true });
+        expect(result.state('count')).to.equal(2);
+        expect(spy.callCount).to.equal(2);
+      });
+    });
+
+    context('updating props', () => {
+
+      it('should call shouldComponentUpdate, componentWillUpdate and componentDidUpdate', () => {
+        const spy = sinon.spy();
+
+        class Foo extends React.Component {
+          constructor(...args) {
+            super(...args);
+            this.state = {
+              foo: 'state',
+            };
+          }
+          componentWillReceiveProps(nextProps, nextContext) {
+            spy('componentWillReceiveProps', this.props, nextProps, nextContext);
+          }
+          shouldComponentUpdate(nextProps, nextState, nextContext) {
+            spy('shouldComponentUpdate', this.props, nextProps, this.state, nextState, nextContext);
+            return true;
+          }
+          componentWillUpdate(nextProps, nextState, nextContext) {
+            spy('componentWillUpdate', this.props, nextProps, this.state, nextState, nextContext);
+          }
+          componentDidUpdate(prevProps, prevState, prevContext) {
+            spy('componentDidUpdate', prevProps, this.props, prevState, this.state, prevContext);
+          }
+          render() {
+            return <div>foo</div>;
+          }
+        }
+        Foo.contextTypes = {
+          foo: React.PropTypes.string,
+        };
+
+        const wrapper = shallow(
+          <Foo foo="bar" />,
+          {
+            context: { foo: 'context' },
+            lifecycleExperimental: true,
+          }
+        );
+        wrapper.setProps({ foo: 'baz' });
+        expect(spy.args).to.deep.equal(
+          [
+            [
+              'shouldComponentUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+            [
+              'componentWillReceiveProps',
```

Comment:
`componentWillReceiveProps` should be called before `shouldComponentUpdate` (see this test for an example https://github.com/facebook/react/blob/c0ecde687ae3e599dced83a111e57ed49897b0b9/src/renderers/shared/stack/reconciler/__tests__/ReactComponentLifeCycle-test.js#L565)